diff --git a/scripts/environments/build.py b/scripts/environments/build.py
index 6d5998b..1ab371b 100644
--- a/scripts/environments/build.py
+++ b/scripts/environments/build.py
@@ -75,17 +75,12 @@ def build_eval_env(
     envs = gym.make(id, num_envs=num_envs, **env_kwargs)
     if isinstance(envs.action_space, gym.spaces.Dict):
         envs = FlattenActionSpaceWrapper(envs)
-    n_steps = kwargs['num_steps']
     if capture_video:
-        run_name = kwargs['run_name']
-        video_dir = f"runs/{run_name}/videos"
-        output_dir = f"{kwargs['checkpoint_dir']}/test_videos"
         print(f"Saving eval videos to {kwargs['checkpoint_dir']}")
-        envs = RecordEpisode(envs, output_dir=os.path.join(output_dir,'test_video'), save_trajectory=True, trajectory_name="trajectory",
-                                  max_steps_per_video=n_steps, video_fps=30)
-
-    envs = ManiSkillVectorEnv(envs, n_steps, ignore_terminations=not kwargs['partial_reset'], **env_kwargs)
+        envs = RecordEpisode(envs, output_dir=f"runs/{kwargs['run_name']}/train_videos", save_trajectory=True, trajectory_name="trajectory",
+                                  max_steps_per_video=kwargs['num_steps'], video_fps=30)
 
+    envs = ManiSkillVectorEnv(envs, num_envs, ignore_terminations=not kwargs['partial_reset'], **env_kwargs)
 
     game = kwargs['game']
     with open(f'{os.path.dirname(os.path.realpath(__file__))}/configuration.yaml', "r") as file:
